vllm/v1/structured_output/backend_guidance.py

```diff
@@ -109,12 +110,39 @@ def accept_tokens(self, request_id: str, tokens: list[int]) -> bool:
         # For conversion between the two, see
         # https://github.com/guidance-ai/llguidance/blob/main/docs/fast_forward.md
 
+        if self.ll_matcher.validate_tokens(tokens) != len(tokens):
+            breakpoint()
+            pass
+
+        self.check_error()
+
         r = self.ll_matcher.consume_tokens(tokens)
 
         self.check_error()
 
         return r
 
+    def validate_tokens(self, tokens: list[int]) -> list[int]:
+        """Checks if the list of tokens are accepted by the parser in sequence.
+        Will not advance the parser.
+
+        Returns the prefix list of tokens that are accepted by the parser.
+        """
+        if len(tokens) == 0:
+            return []
+        if self.ll_matcher.is_stopped():
+            return []
+
+        num_tokens = self.ll_matcher.validate_tokens(tokens)
+
+        self.check_error()
+
+        return tokens[:num_tokens]
+
+    def rollback(self, num_tokens: int) -> None:
+        self.ll_matcher.rollback(num_tokens)
+        self.check_error()
```

Comment:
I came across the [rollback API](https://github.com/guidance-ai/llguidance/issues/117) you requested. Do we need this when we use the above `self.ll_matcher.validate_tokens(tokens)` if it does not forward the state of the parser?

Reply:
I see...the rollback api is also used in `vllm/v1/structured_output/__init__.py` so it cant be removed

Reply:
Yes, this API is needed so that we can advance the state to generate bitmasks for speculated tokens, and then rollback to preserve the initial state.